pkg/generator/password_reconciler.go

```diff
@@ -115,37 +123,96 @@ func (r *PasswordReconciler) createSecret(ctx context.Context, password *sgv1alp
+
+	var password strings.Builder
+
+	//Set symbol character
+	for i := 0; i < spec.Symbols; i++ {
+		value, _ := randomElement(spec.SymbolCharSet)
+		password.WriteString(value)
+	}
+
+	//Set digit
+	for i := 0; i < spec.Digits; i++ {
+		value, _ := randomElement(digitSet)
+		password.WriteString(value)
+	}
+
+	//Set uppercase
+	for i := 0; i < spec.UppercaseLetters; i++ {
+		value, _ := randomElement(upperCharSet)
+		password.WriteString(value)
+	}
 
-	genParams := passwordParams{Length: password.Spec.Length}
-	if genParams.Length == 0 {
-		genParams.Length = 40
+	//Set lowercase
+	for i := 0; i < spec.LowercaseLetters; i++ {
+		value, _ := randomElement(lowerCharSet)
+		password.WriteString(value)
 	}
 
-	passwordVal, err := gen.Generate(genParams)
+	var allCharSet = lowerCharSet + upperCharSet + digitSet
+
+	remainingLength := spec.Length - spec.Symbols - spec.Digits - spec.UppercaseLetters - spec.LowercaseLetters
+	for i := 0; i < remainingLength; i++ {
+		value, _ := randomElement(allCharSet)
+		password.WriteString(value)
+	}
+
+	inRune := []rune(password.String())
+	localCryptoShuffle(len(inRune), func(i, j int) {
+		inRune[i], inRune[j] = inRune[j], inRune[i]
+	})
+
+	return string(inRune)
+}
+
+func randomElement(s string) (string, error) {
+	n, err := rand.Int(rand.Reader, big.NewInt(int64(len(s))))
 	if err != nil {
 		return "", err
 	}
+	return string(s[n.Int64()]), nil
+}
+
+func localCryptoShuffle(n int, swap func(i, j int)) {
+	if n < 0 {
+		panic("invalid argument to Shuffle")
+	}
 
-	return passwordVal.(string), nil
+	// Fisher-Yates shuffle: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+	// Shuffle really ought not be called with n that doesn't fit in 32 bits.
+	// Not only will it take a very long time, but with 2³¹! possible permutations,
+	// there's no way that any PRNG can have a big enough internal state to
+	// generate even a minuscule percentage of the possible permutations.
+	// Nevertheless, the right API signature accepts an int n, so handle it as best we can.
+	i := n - 1
+	for ; i > 1<<31-1-1; i-- {
+		j, _ := rand.Int(rand.Reader, big.NewInt(int64(i+1)))
+		swap(i, int(j.Int64()))
+	}
+	for ; i > 0; i-- {
+		j, _ := rand.Int(rand.Reader, big.NewInt(int64(i+1)))
+		swap(i, int(j.Int64()))
+	}
 }
 
 func (r *PasswordReconciler) updateStatus(ctx context.Context, password *sgv1alpha1.Password) error {
 	existingPassword, err := r.sgClient.SecretgenV1alpha1().Passwords(password.Namespace).Get(ctx, password.Name, metav1.GetOptions{})
 	if err != nil {
-		return fmt.Errorf("Fetching password: %s", err)
+		return fmt.Errorf("fetching password: %s", err)
 	}
 
 	existingPassword.Status = password.Status
 
 	_, err = r.sgClient.SecretgenV1alpha1().Passwords(existingPassword.Namespace).UpdateStatus(ctx, existingPassword, metav1.UpdateOptions{})
 	if err != nil {
-		return fmt.Errorf("Updating password status: %s", err)
+		return fmt.Errorf("updating password status: %s", err)
```

Comment:
Question for other maintainers, i.e @joe-kimmel-vmw / @cppforlife seems like we have a mix of upper and lower case error formatting going on. I think go conventionally pushes for the lower case, but I am see a LOT of upper cased error messaging in this repo. Any thoughts on  why we went with the uppercasing generally?

Reply:
carvel projects always used capital letter for starting errors.

Reply:
> Any thoughts on why we went with the uppercasing generally?

i found it easier to read when k14s was started and also it was easier to split errors to be more human friendly.

Reply:
+1 - it's initial caps because that's what DK prefers.  I think i tried to put that in our mostly-useless style guide at some point.